Cython/Compiler/Pipeline.py

```diff
@@ -288,8 +288,9 @@ def create_pyx_as_pxd_pipeline(context, result):
     for stage in pyx_pipeline:
         pipeline.append(stage)
         if isinstance(stage, AnalyseDeclarationsTransform):
-            # This is the last stage we need.
-            break
+            # crude hack to not generate pickle code for classes cimported
+            stage._inject_pickle_methods = lambda node: None
```

Comment:
Maybe it's just an "extern" missing as visibility somewhere? Because that's what parsing a `.pxd` file should give us, declarations external to the current module.

Reply:
Yes - that's probably right. I'll have a look...

Reply:
It looks like what actually controls pickle function generation is `scope.implemented`. This is set in `CClassDefNode.analyse_declarations` according to whether `in_pxd` is set. I've therefore made a change to set `in_pxd` just prior to the `analyse_declarations` stage.

With that says, it feels like quite a drastic change with the potential to break stuff, so I'm not really convinced its the right thing to do either.

Reply:
I'd rather say this should have been done right from the start, and is the right thing to change in 3.0.

Your change is obviously a quick shot to try it out. In the end, this should be done right after parsing. Ideally, we'd follow the `.pxd` transformations from that point on (such as `PxdPostParse`), but some of it might fail in a `.pyx` parsed tree (which would contain a lot of stuff that is illegal in `.pxd` files). Maybe we could get there by cleaning the tree somehow. Probably not something to do right now.

I think the `in_pxd` attributes only occur on `StatNode`s, not arbitrary `Node`s. Intercepting on those should be enough.

I'm not sure how `inline` functions should behave. The author of a `.pyx` file would probably consider them local, but cimporting for it would allow them to "leak" into other modules. That could lead to problems when the function depends on other globals defined in the `.pyx` module. I'd rather disallow `inline` functions to be cimported from `.pyx` files.

Reply:
Right. For the moment I've just changed `Node` -> `StatNode` and not changed anything else.

As you say, it'd be better to put cimported `.pyx`s through as much of the same pipeline as `.pxd` files as possible. But that probably should be done separately. It possibly might be best to just cimported pyx part of this PR rather than changing this bit twice? I'm not really sure.